### PR TITLE
AWS Lambda response streaming for Lambda URLs

### LINKED ISSUE #11906: AWS Lambda response streaming for function urls
### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Use case description

AWS recently released support for [response streaming](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-response-streaming/) accessible through Lambda Function URLs.

To utilize this new integration, `AWS::Lambda::Url` includes an optional property, `InvokeMode`. When the Lambda Function URL invoke mode is set to `RESPONSE_STREAM`, the Lambda function handler will be invoked using the new InvokeWithResponseStream API.

While Serverless Framework can provision AWS::Lambda::Url, it is not exposing InvokeMode property.

### Proposed solution (optional)

Expose optional `url.invoke: 'buffered' | 'response_stream'` property in Lambda Function URL configuration mapping to [AWS::Lambda::Url InvokeMode property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-url.html).

Shorter name and lowercase values are matching `authorizer` that maps to the AuthType property.

### PR DESCRIPTION
**Pull request by grakic**
Closes: #11906

### CODE REVIEW

**Comment by dested** [comment]
Any thoughts on when/if this will get merged? 

Also any thoughts on if its safe enough to just patch on to my serverless project? It looks relatively simple, just not sure if its missing anything I should be aware of

**Review by GrahamCampbell — DISMISSED** [review]

File: lib/plugins/aws/package/compile/functions.js
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions.js:710]: I think simpler would be to simply contain whole resolution here as:

```javascript
if (url.invoke === 'RESPONSE_STREAM) urlResource.Properties.InvokeMode = url.invoke;
```

File: test/fixtures/programmatic/function/stream.js
Comment (medikoo) [inline test/fixtures/programmatic/function/stream.js:3]: Instead of doing that, let's maybe add `awsLambda` as global into ESLint config, for the lambda fixtures.

ESlint config is maintained in `package.json`

**Review by medikoo — CHANGES_REQUESTED** [review]
@grakic that looks really good! I have just a few mostly cosmetic suggestions

File: package.json
Comment (medikoo) [inline package.json:135]: Let's reuse block at L117 (it's dedicated for fixtures)

File: test/integration/aws/function-url.test.js
Comment (medikoo) [inline test/integration/aws/function-url.test.js:124]: Is it confirmed that test passes?

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @grakic it looks great. I have just one final small organization suggestion and we're ready to go

File: test/integration/aws/function-url.test.js
Comment (harounansari-cj) [reply test/integration/aws/function-url.test.js:124]: Can you guys merge this please, im tryna use it

**Comment by harounansari-cj** [comment]
Hey guys, Im frustrated that we've had lambda response streaming and SAM has had that feature for a while and Serverless Framework has lagged behind.

I could be missing something, is there another way to add response streaming to this function:
```
functions:
  hello:
    handler: handler.hello
    url:
      # Allow CORS for all requests from any origin
      cors: true
```

I can do it in my SAM code like this:
```
  StreamingFunctionUrl:
    Type: AWS::Lambda::Url
    Properties:
      TargetFunctionArn: !Ref StreamingFunction
      AuthType: NONE
      InvokeMode: RESPONSE_STREAM
```

**Comment by dested** [comment]
> Hey guys, Im frustrated that we've had lambda response streaming and SAM has had that feature for a while and Serverless Framework has lagged behind.



I ended up applying a patch using pnpm patching. It's for serverless version 3.30.1 but you should be able to make similar changes in whatever version your using.


serverless@3.30.1.patch:
```
diff --git a/lib/plugins/aws/package/compile/functions.js b/lib/plugins/aws/package/compile/functions.js
index 7de85631abe83463672973bef50f5cb1a264589f..0c0afe5ccc75c2f3039ed0f986f0aa9ed0058762 100644
--- a/lib/plugins/aws/package/compile/functions.js
+++ b/lib/plugins/aws/package/compile/functions.js
@@ -642,6 +642,7 @@ class AwsCompileFunctions {
 
     let auth = 'NONE';
     let cors = null;
+    let invoke=null
     if (url.authorizer === 'aws_iam') {
       auth = 'AWS_IAM';
     }
@@ -675,7 +676,13 @@ class AwsCompileFunctions {
 
       cors.maxAge = url.cors.maxAge;
     }
-
+    if (url.invoke) {
+      if (url.invoke === 'buffered') {
+        invoke = 'BUFFERED';
+      } else if (url.invoke === 'response_stream') {
+        invoke = 'RESPONSE_STREAM';
+      }
+    }
     const urlResource = {
       Type: 'AWS::Lambda::Url',
       Properties: {
@@ -696,6 +703,9 @@ class AwsCompileFunctions {
       };
     }
 
+    if (invoke) {
+      urlResource.Properties.InvokeMode = invoke;
+    }
     const logicalId = this.provider.naming.getLambdaFunctionUrlLogicalId(functionName);
     cfTemplate.Resources[logicalId] = urlResource;
     cfTemplate.Outputs[this.provider.naming.getLambdaFunctionUrlOutputLogicalId(functionName)] = {
diff --git a/lib/plugins/aws/provider.js b/lib/plugins/aws/provider.js
index c737ae300234fe733cdbdcdfcc5297bafe772401..5a8791d2f711b668658724a8086859948d772ca9 100644
--- a/lib/plugins/aws/provider.js
+++ b/lib/plugins/aws/provider.js
@@ -1495,6 +1495,10 @@ class AwsProvider {
                         },
                       ],
                     },
+                    invoke: {
+                      type: 'string',
+                      enum: ['buffered', 'response_stream'],
+                    },
                   },
                   additionalProperties: false,
                 },
```

In your package.json:

```json
  "pnpm": {
    "patchedDependencies": {
      "serverless@3.30.1": "patches/serverless@3.30.1.patch"
    }
  },
```

Best of luck!

**Comment by harounansari-cj** [comment]
> > Hey guys, Im frustrated that we've had lambda response streaming and SAM has had that feature for a while and Serverless Framework has lagged behind.
> 
> I ended up applying a patch using pnpm patching. It's for serverless version 3.30.1 but you should be able to make similar changes in whatever version your using.
> 
> [serverless@3.30.1.patch](mailto:serverless@3.30.1.patch):
> 
> ```
> diff --git a/lib/plugins/aws/package/compile/functions.js b/lib/plugins/aws/package/compile/functions.js
> index 7de85631abe83463672973bef50f5cb1a264589f..0c0afe5ccc75c2f3039ed0f986f0aa9ed0058762 100644
> --- a/lib/plugins/aws/package/compile/functions.js
> +++ b/lib/plugins/aws/package/compile/functions.js
> @@ -642,6 +642,7 @@ class AwsCompileFunctions {
>  
>      let auth = 'NONE';
>      let cors = null;
> +    let invoke=null
>      if (url.authorizer === 'aws_iam') {
>        auth = 'AWS_IAM';
>      }
> @@ -675,7 +676,13 @@ class AwsCompileFunctions {
>  
>        cors.maxAge = url.cors.maxAge;
>      }
> -
> +    if (url.invoke) {
> +      if (url.invoke === 'buffered') {
> +        invoke = 'BUFFERED';
> +      } else if (url.invoke === 'response_stream') {
> +        invoke = 'RESPONSE_STREAM';
> +      }
> +    }
>      const urlResource = {
>        Type: 'AWS::Lambda::Url',
>        Properties: {
> @@ -696,6 +703,9 @@ class AwsCompileFunctions {
>        };
>      }
>  
> +    if (invoke) {
> +      urlResource.Properties.InvokeMode = invoke;
> +    }
>      const logicalId = this.provider.naming.getLambdaFunctionUrlLogicalId(functionName);
>      cfTemplate.Resources[logicalId] = urlResource;
>      cfTemplate.Outputs[this.provider.naming.getLambdaFunctionUrlOutputLogicalId(functionName)] = {
> diff --git a/lib/plugins/aws/provider.js b/lib/plugins/aws/provider.js
> index c737ae300234fe733cdbdcdfcc5297bafe772401..5a8791d2f711b668658724a8086859948d772ca9 100644
> --- a/lib/plugins/aws/provider.js
> +++ b/lib/plugins/aws/provider.js
> @@ -1495,6 +1495,10 @@ class AwsProvider {
>                          },
>                        ],
>                      },
> +                    invoke: {
> +                      type: 'string',
> +                      enum: ['buffered', 'response_stream'],
> +                    },
>                    },
>                    additionalProperties: false,
>                  },
> ```
> 
> In your package.json:
> 
> ```json
>   "pnpm": {
>     "patchedDependencies": {
>       "serverless@3.30.1": "patches/serverless@3.30.1.patch"
>     }
>   },
> ```
> 
> Best of luck!

thank you ill give this a shot

**Comment by harounansari-cj** [comment]
Hmm I tried your approach and added it to my package json. I also tried manually updating my node_modules/.bin/serverless and ran `node_modules/.bin/serverless` but it doesnt seem to be changing the invoke mode:
image

**Comment by medikoo** [comment]
@harounansari-cj  @dested it looks that this PR got abandoned, are you interested in finalizing this work?

File: test/integration/aws/function-url.test.js
Comment (grakic) [reply test/integration/aws/function-url.test.js:124]: @medikoo It works locally. I am waiting for someone to approve running tests in CI:

```
$ AWS_ACCESS_KEY_ID=*** AWS_SECRET_ACCESS_KEY="***" npx mocha test/integration/aws/function-url.test.js 
(node:846961) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)


  test/integration/aws/function-url.test.js
2023-05-27T21:05:08.096Z ℹ serverless:test deploy test-function-2fsph87s (at /tmp/tmpdirs-serverless/a99c/11b324)
    ✔ should return valid response from Lambda URL (1008ms)
    ✔ should return valid response from Lambda URL with authorizer with valid signature (891ms)
    ✔ should return invalid response from Lambda URL with authorizer without passed signature (338ms)
    ✔ should return expected CORS headers from Lambda URL (777ms)
    ✔ should return streaming invoke mode Lambda URL (854ms)
2023-05-27T21:07:00.757Z ℹ serverless:test remove test-function-2fsph87s (at /tmp/tmpdirs-serverless/a99c/11b324)


  5 passing (2m)

```

**Comment by grakic** [comment]
@medikoo Can you review proposed updates and resolve conversations?

File: test/integration/aws/function-url.test.js
Comment (medikoo) [reply test/integration/aws/function-url.test.js:124]: Integration tests are not run on PR's (only after merge's)

**Review by medikoo — APPROVED** [review]
Thank you @grakic, looks great 👍

**Comment by jayarjo** [comment]
So is it `invokeMode` or `invoke`?

**Comment by sainiankit** [comment]
What could be causing:
```
ERROR	TypeError: responseStream.setContentType is not a function
    at Object.<anonymous> (/var/task/dist/server.js:11:20)
    at Generator.next (<anonymous>)
    at /var/task/dist/server.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/var/task/dist/server.js:3:12)
    at /var/task/dist/server.js:10:85
    at /var/task/serverless_sdk/index.js:24:10716
    at /opt/sls-sdk-node/wrapper.js:13368:92
    at Runtime.handler (/opt/sls-sdk-node/wrapper.js:13375:11)
    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1173:29)
```
    
   Lambda Handler looks like:
   
```
    declare const awslambda: any;

exports.handler = awslambda.streamifyResponse(
  async (_event: any, responseStream: any, _context: any) => {
    responseStream.setContentType("text/event-stream");
    setTimeout(() => responseStream.write("data: Hello, \n\n"), 1000);
    setTimeout(() => responseStream.write("data: World!\n\n"), 2000);
    setTimeout(() => responseStream.end(), 3000);
  }
);
```

serverless.yml liiks like:
```
service: streaming-lambda

provider:
  name: aws
  runtime: nodejs18.x
  stage: production
  region: ap-south-1
  memorySize: 256

functions:
  app:
    handler: dist/server.handler
    url:
      cors: true
      invokeMode: RESPONSE_STREAM
    timeout: 60
```

**Comment by grakic** [comment]
@sainiankit It seems that you are using serverless_sdk wrapper for your Lambda function.

I don't think serverless_sdk is supporting stream invoke mode, please see discussion here: https://github.com/serverless/serverless/issues/11906#issuecomment-1514504006

**Comment by sainiankit** [comment]
@grakic Sorry, Please could you explain it a little:
>seems that you are using serverless_sdk wrapper for your Lambda function.

I have serverless.yml defined
I use github actions to call serverless deploy.
Where does serverless_sdk come into action ?

Is there a work around ? 


```

      - name: serverless deploy
        uses: serverless/github-action@master
        with:
          args: deploy
        env:
          SERVERLESS_ACCESS_KEY: ${{ secrets.SERVERLESS_ACCESS_KEY }}
    ```

**Comment by tobilg** [comment]
You could have a look at my working example: https://github.com/tobilg/serverless-duckdb/blob/main/src/functions/streamingQuery.ts

**Pull request merged**
